h2o-core/src/main/java/water/api/GridSearchHandler.java

```diff
@@ -140,11 +142,11 @@ protected void validateHyperParams(P params, Map<String, Object[]> hyperParams)
           break;
         }
       }
-      if (fieldMetadata == null) {
+      if (fieldMetadata == null && !hparam.getKey().equals(CONSTRAINTS)) {
         throw new H2OIllegalArgumentException(hparam.getKey(), "grid",
                                               "Unknown hyper parameter for grid search!");
       }
-      if (!fieldMetadata.is_gridable) {
+      if (!hparam.getKey().equals(CONSTRAINTS) && !fieldMetadata.is_gridable) {
```

Comment:
for simplicity and also to add proper validation of the "constraints", I'd suggest sth like the following:
```
Set<String> allKeys = hyperParams.entrySet();
allKeys.remove(CONSTRAINTS);
for (Map.Entry<String, Object[]> hparam : allKeys) {
    ... // no need to test for CONSTRAINTS here anymore
}
if (hyperParams.contains(CONSTRAINTS)) {
    // now you get nested validation for free
    validateHyperParams(params, hyperParams.get(CONSTRAINTS));
}
```